protocols/imap/src/main/java/org/apache/james/imap/decode/main/DefaultImapDecoder.java

```diff
@@ -61,65 +61,63 @@ public DefaultImapDecoder(StatusResponseFactory responseFactory, ImapCommandPars
 
     @Override
     public ImapMessage decode(ImapRequestLineReader request, ImapSession session) {
-        ImapMessage message;
         try {
-            final Tag tag = request.tag();
-            message = decodeCommandTagged(request, tag, session);
+            Tag tag = request.tag();
+            return decodeCommandTagged(request, tag, session);
         } catch (DecodingException e) {
             LOGGER.debug("Cannot parse tag", e);
-            message = unknownCommand(null, session);
+            return unknownCommand(null, session);
         }
-        return message;
     }
 
     private ImapMessage decodeCommandTagged(ImapRequestLineReader request, Tag tag, ImapSession session) {
-        ImapMessage message;
         LOGGER.debug("Got <tag>: {}", tag);
         try {
-            final String commandName = request.atom();
-            message = decodeCommandNamed(request, tag, commandName, session);
+            String commandName = request.atom();
+            return decodeCommandNamed(request, tag, commandName, session);
         } catch (DecodingException e) {
             LOGGER.debug("Error during initial request parsing", e);
-            message = unknownCommand(tag, session);
+            return unknownCommand(tag, session);
         }
-        return message;
     }
 
     private ImapMessage unknownCommand(Tag tag, ImapSession session) {
-        ImapMessage message;
-        Object c = session.getAttribute(INVALID_COMMAND_COUNT);
-        int count = 0;
-        if (c != null) {
-            count = (Integer) c;
-        }
-        count++;
+        int count = retrieveUnknownCommandCount(session);
+
         if (count > maxInvalidCommands || session.getState() == ImapSessionState.NON_AUTHENTICATED) {
-            message = responseFactory.bye(HumanReadableText.BYE_UNKNOWN_COMMAND);
+            ImapMessage message = responseFactory.bye(HumanReadableText.BYE_UNKNOWN_COMMAND);
             session.logout();
+            return message;
         } else {
             session.setAttribute(INVALID_COMMAND_COUNT, count);
             if (tag == null) {
-                message = responseFactory.untaggedBad(HumanReadableText.UNKNOWN_COMMAND);
+                return responseFactory.untaggedBad(HumanReadableText.UNKNOWN_COMMAND);
             } else {
-                message = responseFactory.taggedBad(tag, null, HumanReadableText.UNKNOWN_COMMAND);
+                return responseFactory.taggedBad(tag, null, HumanReadableText.UNKNOWN_COMMAND);
             }
-
         }
+    }
 
-        return message;
+    private int retrieveUnknownCommandCount(ImapSession session) {
+        Object c = session.getAttribute(INVALID_COMMAND_COUNT);
+        int count = 0;
+        if (c != null) {
+            count = (Integer) c;
+        }
+        count++;
+        return count;
```

Comment:
```
if (c == null) {
  return 1;
}
return ((Integer) c) + 1;
```
?

Reply:
That's not tested. I belive mutting directly the session integer count be broken.

I would rather do a get then set explicitly.

IMO any further refactoring (ie more than ctrl + maj +m) needs tests first.

Reply:
oops, ok (I expect the attribute to be set at a point, but ok to be defensive on that)

Reply:
set is done by the caller, the mutating ++ is misleading...